### ISSUE TITLE
WCAG 2.4.9 test 1: false positives

### ISSUE THREAD

**Issue by jzuidweg** [veraPDF/veraPDF-wcag-algs#85]
I'm attaching two files which fail **WCAG 2.4.9 test 1** ('_Missing alternate Link description_'), but should not show this error:

1. The [first file](https://github.com/veraPDF/veraPDF-wcag-algs/files/7374310/mail-link-met-alt-tekst.pdf) contains a link which does have an alternative text tagged as such.
2. The [second file](https://github.com/veraPDF/veraPDF-wcag-algs/files/7374318/mail-link-met-expansie-tekst.pdf) has an expansion text. Not sure what 'expansion text' means but perhaps you can check in the tag structure.

Can you look into these apparent false positives?

**Comment by jzuidweg** [comment]
Thanks for sorting this out! I agree it's a good idea to take into account any of the `/Alt` , `/Contents` elements or text under the link rectangle. Sometimes PDF/UA appears to be stricter than WCAG 2.1 and perhaps too strict for our purpose.

**Comment by GehPlatVwaard** [comment]
If you take the text under the link rectangle into account you will never find an error? So you can switch of the test? Could you replace it with a warning “is the link text descriptive?”.

**Comment by jzuidweg** [comment]
Hi Stephan, good to see you here as well :-)

I'll consider making this a warning. For your information, pdfchecker.nl is made up of a front-end and back-end which are implemented independently. The front-end is implemented by Pleio and the code is also [open source on Gitlab](https://gitlab.com/pleio/pdfchecker).

We don't remove tests here at the back-end. Instead, we can ignore errors at the front-end or categorize them as warnings there. This way we keep the test suite on the back-end stable. The back-end has a full implementation of the Matterhorn protocol which we don't want to mess with as it's a formal and well documented test suite. 

In addition the back-end has a few more experimental and heuristic tests for some WCAG 2.1 criteria. The test this issue refers to (WCAG 2.4.9 test 1) is one of these more experimental tests. Instead of taking it out of the test suite, we'll try to improve it. Meanwhile, if it gives too many false positives or negatives, we can choose to categorise it as a warning (or ignore it completely) in the front-end. This way it won't contribute to the accessibility score either.

If you want to know more about how pdfchecker.nl is implemented, don't hesitate to contact me. You know where to find me :-)

**Comment by bdoubrov** [comment]
I think we need to be more careful disabling such test. We have now multiple choices where alternate text may come from:

- /Contents entry in the Link annotation itself -> this is an official PDF/UA-1 requirement
- /Alt entry in the Link structure element
- text under the link annotation rectangle

Each of these may or may not be meaningful. And in fact the only logic we have now is that we don't treat the text under the link rectangle (the 3rd case) as meaningful if it is a URL (starts with http:// or https://). 

So, there are multiple scenarios possible here in case of WCAG. I'd say, that the most logical would be to try retrieving the alternate link description from any of the above 3 sources and then checking that it is non-empty and does not start with http/https.

Alternatively, we can have a logic closer to PDF/UA requirements: check that either /Alt or /Contents entries are non-empty. 

But if we disable this check completely, we can easily skip the accessibility issue when we have a button (no text, just an image or a line art) that works as a link and has no alternative descriptions at all.

**Comment by GehPlatVwaard** [comment]
I agree not to disable the check completely and indeed  the most logical would be your proposal:  retrieving the alternate link description from any of the above 3 sources and then checking that it is non-empty and does not start with http/https.
